keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1604 @@
+wrapper) in the future if that seemed like a better idea. (In theory
+we could also use netlink directly, without needing cgo or external
+libraries, but this would probably be a bad idea; `libnftables`
+implements quite a bit of functionality on top of the raw netlink
+API.)
+
+When reading data from the kernel (`nft list ...`), `nft` outputs the
+data in a nested "object" form:
+
+```
+table ip kube_proxy {
+  comment "Kubernetes service proxying rules";
+
+  chain services {
+    ip daddr . ip protocol . th dport vmap @service_ips
+  }
+}
+```
+
+When writing data to the kernel, `nft` accepts the data in either the
+same "object" form used by `nft list`, or in the form of a set of
+`nft` command lines without the leading "`nft`" (which are then
+executed atomically):
+
+```
+add table ip kube_proxy { comment "Kubernetes service proxying rules"; }
+add chain ip kube_proxy services
+add rule ip kube_proxy services ip daddr . ip protocol . th dport vmap @service_ips
+```
+
+The "object" form is more logical and easy to understand, but the
+"command" form is better for dynamic usage. In particular, it allows
+you to add and remove individual chains, rules, map/set elements, etc,
+without needing to also include the chains/rules/elements that you are
+not modifying.
+
+The examples below all show the "object" form of data, but it should
+be understood that these are examples of what would be seen in `nft
+list` output after kube-proxy creates the rules (with additional
+`#`-preceded comments added to help the KEP reader), not examples of
+the data we will actually be passing to `nft`.
+
+The examples below are also all IPv4-specific, for simplicity. When
+actually writing out rules for nft, we will need to switch between,
+e.g., "`ip daddr`" and "`ip6 daddr`" appropriately, to match an IPv4
+or IPv6 destination address. This will actually be fairly simple
+because the `nft` command lets you create "variables" (really
+constants) and substitute their values into the rules. Thus, we can
+just always have the rule-generating code write "`$IP daddr`", and
+then pass either "`-D IP=ip`" or "`-D IP=ip6`" to `nft` to fix it up.)
+
+(Also, most of the examples below have not actually been tested and
+may have syntax errors. Caveat lector.)
+
+`nft` also has a JSON mode which theoretically should be easier to use
+(both in terms of parsing `nft` output and in terms of not having to
+worry as much about the quoting of input). However, the JSON syntax of
+individual chain rules is poorly documented, and often _extremely_
+different from the non-JSON syntax. (In general, the only way to
+reliably figure out what the JSON syntax is for a particular rule is
+to get `nft` to translate it for you.) So specifying rules this way
+would be harder to understand and maintain. For now, at least, the
+plan is to always use the non-JSON syntax, at least for anything
+involving rules or set/map elements.
+
```

Comment:
What stability guarantees are available for the non-JSON interface? Do we want to look these up before beta?

Reply:
Yeah, I wondered if I should update this section some more...

I have been talking with one of the nftables developers. The CLI API is considered "stable", but without strong ideas about exactly what that means (eg, right now the output has whitespace both before and after every kind of punctuation except commas, which have whitespace after but not before... Is that guaranteed? :man_shrugging:). He's going to bring it up at the [Netfilter Workshop](https://workshop.netfilter.org/2023/) in a few weeks.

I also realized while talking with him that we can probably get away with using the non-JSON API for the "write" commands (`add`, `delete`, `flush`, `insert`, `replace`), but use the JSON API for the "read" commands (`list`, `get`), because for kube-proxy's purposes, we never actually need to read back a `rule` we previously created, and for every type other than `rule`, we can parse the JSON responses correctly. (The JSON syntax for map/set elements is _annoying_, but not actually unusable.)

Reply:
Did you ask about the `libnftables` API guarantees?

Reply:
The Netfilter Workshop was the last two days... no word yet on what came out of discussions. But I've already started implementing what I said in the comment above; we can use the JSON mode for things we need to parse the results of, because for our purposes we don't need to be able to parse back rules we previously created.